File: Server~/src/tools/scripting/UpdateScriptTool.ts
import { z } from 'zod';
import { BaseTool } from '../base/BaseTool.js';
import { Tool, Tags, Examples } from '../base/ToolDecorators.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const UpdateScriptArgsSchema = z.object({
  scriptPath: z.string().describe('Path to the script file (e.g., Assets/Scripts/PlayerController.cs)'),
  operation: z.enum(['add_field', 'add_method', 'add_using', 'rename_class', 'add_namespace', 'remove_field', 'remove_method'])
    .describe('Type of update operation'),
  field: z.object({
    name: z.string(),
    type: z.string(),
    isPublic: z.boolean().default(false),
    isSerializeField: z.boolean().default(true),
    defaultValue: z.string().optional()
  }).optional().describe('Field to add (for add_field operation)'),
  method: z.object({
    name: z.string(),
    returnType: z.string().default('void'),
    parameters: z.array(z.object({
      name: z.string(),
      type: z.string()
    })).optional(),
    body: z.string().optional(),
    isPublic: z.boolean().default(true),
    isStatic: z.boolean().default(false)
  }).optional().describe('Method to add (for add_method operation)'),
  usingStatement: z.string().optional().describe('Using statement to add (for add_using operation)'),
  newClassName: z.string().optional().describe('New class name (for rename_class operation)'),
  namespace: z.string().optional().describe('Namespace to add (for add_namespace operation)'),
  targetName: z.string().optional().describe('Name of field/method to remove (for remove operations)')
});

@Tool({
  name: 'update_script',
  description: 'Update existing C# scripts by adding/removing fields, methods, using statements, or renaming classes',
  category: 'scripting',
  version: '1.0.0'
})
@Tags(['unity', 'scripting', 'csharp', 'refactoring', 'code-modification'])
@Examples([
  {
    description: 'Add a field to existing script',
    args: {
      scriptPath: 'Assets/Scripts/PlayerController.cs',
      operation: 'add_field',
      field: {
        name: 'health',
        type: 'int',
        isPublic: false,
        isSerializeField: true,
        defaultValue: '100'
      }
    }
  },
  {
    description: 'Add a method to existing script',
    args: {
      scriptPath: 'Assets/Scripts/PlayerController.cs',
      operation: 'add_method',
      method: {
        name: 'TakeDamage',
        returnType: 'void',
        parameters: [{ name: 'amount', type: 'int' }],
        body: 'health -= amount;',
        isPublic: true
      }
    }
  },
  {
    description: 'Add using statement',
    args: {
      scriptPath: 'Assets/Scripts/PlayerController.cs',
      operation: 'add_using',
      usingStatement: 'UnityEngine.UI'
    }
  },
  {
    description: 'Rename class',
    args: {
      scriptPath: 'Assets/Scripts/Player.cs',
      operation: 'rename_class',
      newClassName: 'PlayerCharacter'
    }
  }
])
export class UpdateScriptTool extends BaseTool {
  get name() { return 'update_script'; }
  get description() { return 'Update existing C# scripts by adding/removing fields, methods, using statements, or renaming classes'; }
  get inputSchema() { return UpdateScriptArgsSchema; }
  get category() { return 'scripting'; }

  protected async beforeExecute(args: any): Promise<void> {
    if (!args.scriptPath || !args.scriptPath.endsWith('.cs')) {
      throw new Error('Script path must end with .cs extension');
    }
    if (!args.scriptPath.startsWith('Assets/')) {
      throw new Error('Script path must start with "Assets/"');
    }

    // Validate operation-specific requirements
    switch (args.operation) {
      case 'add_field':
        if (!args.field) {
          throw new Error('field parameter is required for add_field operation');
        }
        if (!/^[a-z][a-zA-Z0-9]*$/.test(args.field.name)) {
          throw new Error('Field name must start with lowercase letter and contain only alphanumeric characters');
        }
        break;
      case 'add_method':
        if (!args.method) {
          throw new Error('method parameter is required for add_method operation');
        }
        if (!/^[A-Z][a-zA-Z0-9]*$/.test(args.method.name)) {
          throw new Error('Method name must start with uppercase letter and contain only alphanumeric characters');
        }
        break;
      case 'add_using':
        if (!args.usingStatement) {
          throw new Error('usingStatement parameter is required for add_using operation');
        }
        break;
      case 'rename_class':
        if (!args.newClassName) {
          throw new Error('newClassName parameter is required for rename_class operation');
        }
        if (!/^[A-Z][a-zA-Z0-9]*$/.test(args.newClassName)) {
          throw new Error('Class name must start with uppercase letter and contain only alphanumeric characters');
        }
        break;
      case 'add_namespace':
        if (!args.namespace) {
          throw new Error('namespace parameter is required for add_namespace operation');
        }
        break;
      case 'remove_field':
      case 'remove_method':
        if (!args.targetName) {
          throw new Error('targetName parameter is required for remove operations');
        }
        break;
    }
  }

  protected formatSuccessResponse(result: any): CallToolResult {
    const { scriptPath, operation, modifiedContent } = result;
    let operationText = '';
    
    switch (operation) {
      case 'add_field': operationText = 'Added field'; break;
      case 'add_method': operationText = 'Added method'; break;
      case 'add_using': operationText = 'Added using statement'; break;
      case 'rename_class': operationText = 'Renamed class'; break;
      case 'add_namespace': operationText = 'Added namespace'; break;
      case 'remove_field': operationText = 'Removed field'; break;
      case 'remove_method': operationText = 'Removed method'; break;
    }

    return {
      content: [{
        type: 'text',
        text: `Success! ${operationText}\nScript: ${scriptPath}\n\nTip: Unity will automatically recompile the script.`
      }]
    };
  }

  protected formatErrorResponse(error: any): CallToolResult {
    const errorMessage = error.message || 'Unknown error occurred';
    let helpText = '';
    
    if (errorMessage.includes('not found')) {
      helpText = '\n\nTip: Make sure the script file exists at the specified path.';
    } else if (errorMessage.includes('already exists')) {
      helpText = '\n\nTip: The field or method you are trying to add already exists in the script.';
    } else if (errorMessage.includes('syntax error')) {
      helpText = '\n\nTip: The modification would create invalid C# syntax. Check your parameters.';
    }
    
    return {
      content: [{ type: 'text', text: `Error updating script: ${errorMessage}${helpText}` }],
      isError: true
    };
  }
}

File: Editor/Tools/Scripting/UpdateScriptTool.cs
using System;
using System.IO;
using System.Text.RegularExpressions;
using UnityEngine;
using UnityEditor;
using McpUnity.Utils;
using McpUnity.Unity;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Update existing C# scripts by adding/removing fields, methods, using statements, or renaming classes
    /// Note: This is a basic implementation. Full refactoring requires Roslyn integration.
    /// Unity API: File system operations
    /// </summary>
    public class UpdateScriptTool : McpToolBase
    {
        public UpdateScriptTool()
        {
            Name = "update_script";
            Description = "Update existing C# scripts by adding/removing fields, methods, using statements, or renaming classes";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            try
            {
                string scriptPath = parameters["scriptPath"]?.ToString();
                string operation = parameters["operation"]?.ToString();
                
                if (string.IsNullOrEmpty(scriptPath))
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        "scriptPath is required", "validation_error");
                }

                if (string.IsNullOrEmpty(operation))
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        "operation is required (add_field, add_method, add_using, rename_class, etc.)", "validation_error");
                }

                // Support relative paths starting with Assets/
                string fullPath = scriptPath;
                if (scriptPath.StartsWith("Assets/") || scriptPath.StartsWith("Assets\\"))
                {
                    fullPath = Path.Combine(Application.dataPath.Replace("/Assets", ""), scriptPath.Replace("/", Path.DirectorySeparatorChar.ToString()));
                }

                if (!File.Exists(fullPath))
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        $"Script file not found: {scriptPath}. Full path checked: {fullPath}", "not_found");
                }

                string content = File.ReadAllText(fullPath);
                string newContent = content;

                switch (operation)
                {
                    case "add_field":
                        newContent = AddField(content, parameters);
                        break;

                    case "add_method":
                        newContent = AddMethod(content, parameters);
                        break;

                    case "add_using":
                        newContent = AddUsing(content, parameters);
                        break;

                    case "rename_class":
                        newContent = RenameClass(content, scriptPath, parameters);
                        break;

                    case "add_namespace":
                        newContent = AddNamespace(content, parameters);
                        break;

                    case "remove_field":
                        newContent = RemoveField(content, parameters);
                        break;

                    case "remove_method":
                        newContent = RemoveMethod(content, parameters);
                        break;

                    default:
                        return McpUnitySocketHandler.CreateErrorResponse(
                            $"Unknown operation: {operation}", "validation_error");
                }

                bool wasModified = content != newContent;
                
                if (wasModified)
                {
                    File.WriteAllText(fullPath, newContent);
                    AssetDatabase.Refresh();
                    McpLogger.LogInfo($"Updated script: {scriptPath} (operation: {operation})");
                }
                else
                {
                    McpLogger.LogInfo($"Script unchanged: {scriptPath} (operation: {operation}, skipped - already exists or no change needed)");
                }

                return new JObject
                {
                    ["success"] = true,
                    ["message"] = wasModified ? "Script updated successfully" : "Script unchanged (item already exists or no change needed)",
                    ["scriptPath"] = scriptPath,
                    ["operation"] = operation,
                    ["modified"] = wasModified
                };
            }
            catch (Exception ex)
            {
                McpLogger.LogError($"UpdateScriptTool error: {ex.Message}");
                return McpUnitySocketHandler.CreateErrorResponse($"Failed: {ex.Message}", "execution_error");
            }
        }

        private string AddField(string content, JObject parameters)
        {
            var field = parameters["field"];
            if (field == null) return content;

            string fieldName = field["name"]?.ToString();
            string fieldType = field["type"]?.ToString();
            string access = field["access"]?.ToString() ?? "public";
            string defaultValue = field["defaultValue"]?.ToString();

            if (string.IsNullOrEmpty(fieldName))
            {
                throw new ArgumentException("Field name is required");
            }

            // Check if field already exists (prevents CS0102 duplicate definition error)
            string fieldPattern = $@"\b{Regex.Escape(fieldName)}\s*[;=]";
            if (Regex.IsMatch(content, fieldPattern))
            {
                McpLogger.LogWarning($"Field '{fieldName}' already exists in script, skipping add_field");
                return content; // Field already exists, don't add duplicate
            }

            string fieldDeclaration = $"    {access} {fieldType} {fieldName}";
            if (!string.IsNullOrEmpty(defaultValue))
            {
                fieldDeclaration += $" = {defaultValue}";
            }
            fieldDeclaration += ";";

            // Insert after class declaration
            int classIndex = content.IndexOf("{", content.IndexOf("class "));
            if (classIndex > 0)
            {
                content = content.Insert(classIndex + 1, "\n" + fieldDeclaration + "\n");
            }

            return content;
        }

        private string AddMethod(string content, JObject parameters)
        {
            var method = parameters["method"];
            if (method == null) return content;

            string methodName = method["name"]?.ToString();
            string returnType = method["returnType"]?.ToString() ?? "void";
            string access = method["access"]?.ToString() ?? "public";
            string body = method["body"]?.ToString() ?? "        // TODO: Implement method";

            if (string.IsNullOrEmpty(methodName))
            {
                throw new ArgumentException("Method name is required");
            }

            // Check if method already exists (prevents duplicate method error)
            string methodPattern = $@"\b{Regex.Escape(methodName)}\s*\(";
            if (Regex.IsMatch(content, methodPattern))
            {
                McpLogger.LogWarning($"Method '{methodName}' already exists in script, skipping add_method");
                return content; // Method already exists, don't add duplicate
            }

            string methodDeclaration = $"    {access} {returnType} {methodName}()\n    {{\n{body}\n    }}\n";

            // Insert before last closing brace
            int lastBrace = content.LastIndexOf("}");
            if (lastBrace > 0)
            {
                content = content.Insert(lastBrace, "\n" + methodDeclaration);
            }

            return content;
        }

        private string AddUsing(string content, JObject parameters)
        {
            string usingStatement = parameters["usingStatement"]?.ToString();
            if (string.IsNullOrEmpty(usingStatement)) return content;

            // Check if using already exists
            if (content.Contains($"using {usingStatement};"))
            {
                return content;
            }

            // Add after other using statements
            int lastUsing = content.LastIndexOf("using ");
            if (lastUsing > 0)
            {
                int lineEnd = content.IndexOf("\n", lastUsing);
                if (lineEnd > 0)
                {
                    content = content.Insert(lineEnd + 1, $"using {usingStatement};\n");
                }
            }
            else
            {
                // Add at the beginning
                content = $"using {usingStatement};\n" + content;
            }

            return content;
        }

        private string RenameClass(string content, string scriptPath, JObject parameters)
        {
            string newClassName = parameters["newClassName"]?.ToString();
            if (string.IsNullOrEmpty(newClassName)) return content;

            // Find current class name
            Match match = Regex.Match(content, @"class\s+(\w+)");
            if (match.Success)
            {
                string oldClassName = match.Groups[1].Value;
                content = content.Replace($"class {oldClassName}", $"class {newClassName}");

                // Rename file
                string directory = Path.GetDirectoryName(scriptPath);
                string newPath = Path.Combine(directory, newClassName + ".cs");
                if (File.Exists(scriptPath) && scriptPath != newPath)
                {
                    File.Move(scriptPath, newPath);
                }
            }

            return content;
        }

        private string AddNamespace(string content, JObject parameters)
        {
            string namespaceName = parameters["namespace"]?.ToString();
            if (string.IsNullOrEmpty(namespaceName)) return content;

            // Wrap content in namespace
            if (!content.Contains($"namespace {namespaceName}"))
            {
                // Find where to start namespace (after usings)
                int insertIndex = 0;
                int lastUsing = content.LastIndexOf("using ");
                if (lastUsing > 0)
                {
                    insertIndex = content.IndexOf("\n", lastUsing) + 1;
                }

                content = content.Insert(insertIndex, $"\nnamespace {namespaceName}\n{{\n");
                content += "\n}";
            }

            return content;
        }

        private string RemoveField(string content, JObject parameters)
        {
            string targetName = parameters["targetName"]?.ToString();
            if (string.IsNullOrEmpty(targetName)) return content;

            // Simple removal - find and remove the line
            string pattern = $@".*{targetName}.*;";
            content = Regex.Replace(content, pattern, "");

            return content;
        }

        private string RemoveMethod(string content, JObject parameters)
        {
            string targetName = parameters["targetName"]?.ToString();
            if (string.IsNullOrEmpty(targetName)) return content;

            // Find method and remove it (basic implementation)
            string pattern = $@"\s*\w+\s+\w+\s+{targetName}\([^)]*\)\s*\{{[^}}]*\}}";
            content = Regex.Replace(content, pattern, "", RegexOptions.Singleline);

            return content;
        }
    }
}
